### ISSUE TITLE
Non-local control of lights

### ISSUE THREAD

**Issue by dudeofawesome** [eliotstocker/Light-Controller#7]
One way to go about this would be to retrieve the user's external IP address when they attempt to control the light, and store that address.  Then, when the user is not connected to their home wifi network, we could switch the app to sending the UDP packets to the external IP address, assuming the user has forwarded the correct port to the bridge.

**Comment by gdbgeek** [comment]
Great thread guys. Keep up the good work.

**Comment by mrwhale** [comment]
@LimitessLED 
Hey Hamish, 
You've implemented this into your version of the android app? Could you comment on this, see if we are on the right track with this feature?

Obviously we have just tried to reverse engineer the commands etc but if we can have comments from someone who has actually implemented this, that would be awesome!

**Comment by TameOfGroans** [comment]
Is this not an ideal use case for VPN? 

Root not-required

https://code.google.com/p/ics-openvpn/

pfSense openvpn (for example)

**Comment by TameOfGroans** [comment]
please not upnp.  instead how about dnssd or mdns?  [zeroconf]

rfc linkage

http://www.zeroconf.org/

**Comment by TameOfGroans** [comment]
Ugh a pox upon www, http://no-www.org , I intended to http://zeroconf.org/

**Comment by LimitlessLED** [comment]
Installing vpn software is too annoying. I would not recommend using it.  And does vpn software support UDP broadcast?  Not normally.

**Comment by LimitlessLED** [comment]
Interesting concept, dnssd or mdns.  Do you have more details on how this could be implemented. I could look at adding it to the wifi controller v5

 

-Hamish

LimitlessLED.

**Comment by mrwhale** [comment]
I don't think any of these are worth implementing. To much overhead, protocols involved. Creating and maintaining a server in the middle (like what is currently implemented) or like what easy bulb is doing with their ones (http://easybulb.com/api/, but add more security) would be the most stable, user friendly. Make the wifi box have a persistant connection to this server, so when you are out and about, you can control the lights by sending commands to this server. If you have a sign up section, so people have to login this would keep it secure. 

I will assume this is how most of the other major smart lights do it too (Phillips hue, belkin) as you have to sign up/login to the apps etc.
Could always make it optional too so its not necessary to login (if you don't want this feature)

Anyways, just my 2 cents on the matter

**Comment by eliotstocker** [comment]
It seems to me were still best trying to implement the built in global control solution, the problem I have is that for whatever reason it doesn't appear to be ring itself up properly at my house, other upnp devices are fine but the box doesn't seem to be opening any ports or anything. And isn't control able from outside the house with the original app. 

So of course with that in mind I'm not quite sure how to go forward with the development on this, maybe I'll put together a quick test that someone can try and we can go from there?

**Comment by mrwhale** [comment]
Sounds good! The upnp works fine in my setup with the original app, so happy to try and tests you have

**Comment by mnwalker** [comment]
Hi,
This is a task I was planning to achieve for my lightbulbs but thinking its probably best to collaborate than compete.

My plan had been to introduce a bunch of additional functionality on a raspberry pi based server.
I know it isn't an out of the box solution, but I simply can't think of any alternative to deliver the feature set I am after.

1) Remote access API (Pi can control DDNS and UPNP, or even initiate a socket connection to central server)
2) IFTTT support (proper channel would be amazing, but until then have workarounds using wordpress calls)
3) Timers and schedules that don't depend on the phone being on the network. (Security lights, Nasa style sleep patterns etc)

For simplest usage this would need a communal plug and play central server, but if you can put a custom address into the app and Pi then those with the knowledge and desire could run their own.

Whats all of your thoughts on this?

Mark

**Comment by mrwhale** [comment]
Hey Mark,
Great idea's you have, I also have a raspberry pi, and do similar things to get around the shortfalls of the wifi lights design

1) Remote access API would be awesome, you could set one up yourself on the pi using (https://github.com/mrwhale/LimitlessLED) this PHP API. Create a php page that is remotely accessible from your phone etc
The remote access server that the wifi box currently uses doesn't have a documented API (apart from what we have found in this thread) so unless someone here can create/maintain a central server, like what easybulb do, everyone is on their own :( (I would be more then happy to maintain such a thing, but this would require someone to find/modify the current firmware to change where the wifibox makes the connection to) You could definitely use the RP to keep initiating the socket request to the current server using cron to periodically contact the server to make sure the connection is still alive

2) I have yet to try out the IFTTT wordpress hack, I am secretly waiting/hoping someone creates an official channel. But unless a proper remote accss API is set up, I don't see this happening

3) you can use this bash script I wrote to control the lights on the raspberry pi (https://github.com/mrwhale/limitlessled-bash) to create timers and schedules with cron. I currently use this in my home setup. i.e. cron will turn the lights on at 6am duing the working week (which is the time i wake up) and it will turn the lights off again at 7am (when I leave)
You could also use this and create schedules. i.e. create a script that checks to see if you phone is on the wifi (is at home) and when you arrive home, turn the lights on ( I would recommend you use Tasker on your phone for this though) But possibilities are there. Let me know what else you are thinking, as I have created a few hacks of my own to use on my pi at home - be happy to share with you

**Comment by LimitlessLED** [comment]
Thanks I have taken those ideas on board,  and may look to include all of them into the limitlessled wifi bridge v5.0 firmware.

 

-          Cloud service address setting via API

-          IFTTT support

-          Timers and Scheduler built into the wifi bridge firmware – settings pushed out to the wifi bridge via API

**Comment by mnwalker** [comment]
Ah great news, having it natively on the controller would be the ideal
solution.
Would eliminate my need for a Pi in the middle and bring costs down.
I can do web and mobile but wouldn't know where to start on the controllers
firmware so glad you are able to help on that side.

On a related note, is the wifi controller for the LimitlessLED much
different compared to the generic milight controllers?
And if different is it in the hardware or just firmware.

Also, IFTTT support may be overkill for the actual controller since it
would need inbuilt DDNS to make that work.
But if you can build Cloud service address setting via API into the
controller then IFTTT can easily be provided.
I would be happy to build the open source cloud service part and host a
communal server for people wanting plug and play ability - could be cross
compatible across the various different apps.
But will need some agreement on the best route and how/where to implement
the IP updates and server calls.


Regards,

*Mark Walker*

*T:* +66 (0) 815 884 380
*E:* m@awcode.com
*W:* http://awcode.com

**Comment by eliotstocker** [comment]
@LimitlessLED Statefullness would be a great idea also, so we could ask the bridge what the last recorded state was :D

**Comment by LimitlessLED** [comment]
That still would not cover the fact that remotes can alter the state of the light too.  And resetting the wifi bridge would lose the state..  it is not possible sorry. They are more like remote command operated only, you just set the light to what you like.

-Hamish.
LimitlessLED.

**Comment by zenitraM** [comment]
Hey, just for anyone who got here looking (as me) for documentation on this "cloud protocol": I have a Milight with its WiFi bridge from a random Aliexpress seller.
I tried using the LimitlessLED Windows control app (works fine on Wine) at http://www.limitlessled.com/dev/

It has some settings to get and set the Cloud API port and host to use:

- `AT+SOCKB` to get it,
- `AT+SOCKB=TCP,38899,www.anymilight.com` to set it.

Someone mentioned earlier the EasyBulb API, at http://easybulb.com/api/. Well, the EasyBulbs seem to be plain Milights, so I figured out you can just point the bulbs to port 38889 on `api.easybulb.com` and then use their HTTP API to remotely interact with your bulbs with your own Mac addresses.

It seems that there is no UPNP or any other trickery on the bridge - just a plain, long-living TCP connection. The "Cloud API" does the rest.

Of course I don't like the idea of using any manufacturer/distributor hosted API - the way should be either finding documentation or reversing the TCP protocol and then creating a proxy which acts as a middleman - I might do this at some point.

**Comment by gdbgeek** [comment]
zenitraM, that is very interesting. I assume your "WiFi bridge from a random Aliexpress seller" must be the V5 bridge as I dont believe the old V4 bridges support those command sets.

**Comment by zenitraM** [comment]
Well, I got it back in April, and there are no markings on it telling the version, it's this one:
http://aliexpress.com/item/2-4G-MI-Light-Wireless-WIFI-RGB-color-Strip-bulb-lamp-internet-Remote-Controller-for-Iphone/2032002345.html?isOrig=true#extend
so not sure if it's a 5.0 or not as the LimitlessLED docs say it was released on September, maybe it's even a 4.0.

Just figured out the "protocol", it's not even a real protocol: the bridge opens a TCP socket to the "cloud server" and sends its MAC as raw text, and periodically sends `Li_Link\n`.

The only thing the server ever sends are the same hex commands that are sent over UDP, when they are sent. So this makes it quite easy to create an "online hub".

**Comment by gdbgeek** [comment]
Awesome, I've got a spare hub so I might try this out, I thought the new commands required a V5 wifi bridge but perhaps I'm mistaken.

**Comment by gdbgeek** [comment]
zenitraM, I can also set the cloud server on my V4 bridge. The easybulb API is really easy to use. Thanks for sharing.

**Comment by LimitlessLED** [comment]
Yes, all bridges versions support those command sets in some capacity or another.

 

Regards,

Hamish.

-LimitlessLED

**Comment by cloudsyjared** [comment]
@gdbgeek @zenitraM are you two having any luck with either of the POST commands (colour/brightness) ? I am getting error 500 for color and invalid JSON for brightness.

**Comment by gdbgeek** [comment]
@cloudsyjared I haven't tried any of the colour/brightness commands. What are you using to send the commands?

**Comment by gdbgeek** [comment]
Actually @cloudsyjared I just tried and also receive a "<h1>Server Error (500)</h1>"

**Comment by cloudsyjared** [comment]
@gdbgeek Firefox plugin

**Comment by gdbgeek** [comment]
@cloudsyjared with the brightness I do get this message

{"success": false, "error": "Device not found"}

Which is probably correct since my bridge isn't on right now.

**Comment by gdbgeek** [comment]
Slightly off topic. @cloudsyjared Im having a decent play with the API now as Im building an intermediate server in order to send commands via IFTTT. I've noticed that I can turn all of my white bulbs on/off but it doesnt work with a specific group. The Easybulb API tells me its successful but no lights are turned on.

**Comment by cloudsyjared** [comment]
@gdbgeek yeah I gave up on the easybulb route and wrote my own this weekend

**Comment by gdbgeek** [comment]
@cloudsyjared your own API? I'm contemplating the same now.

**Comment by cloudsyjared** [comment]
@gdbgeek yep wasn't too bad

**Comment by gdbgeek** [comment]
@zenitraM do you recall in your tests if the MAC address you see is the same MAC as you see in the official app? I'm seeing a similar MAC but the last octet is incremented by 1 (e.g. if the bridge ends with 9E I'm seeing 9F being sent to the server). I've tested 4 bridges and they all do the same thing.

**Comment by eliotstocker** [comment]
I'm thinking we should start a wiki here to keep all of this info in a nice neat order

**Comment by Zulex** [comment]
Is this feature going to be implemented in the app since some of you guys figured out how to control it from other networks. The app from : http://mi-light.com ( not the official milight manufacturer ) is able to do it without any port forwarding or changing settings. Even using colors, someone figured out how? I got the same server error.

**Comment by cloudsyjared** [comment]
I have written an api + service for this that I plan on announcing soon. You can check for updates on the mithings project under my github page

**Comment by mrwhale** [comment]
Hey guys, I think once we have launched milestone 1.8 this will be the first feature to be done for the next milestone. Once I've finished the pebble feature, I might start hacking something together to get a working prototype of this feature, see how easy it's going to be.

Nice stuff @cloudsyjared will definitely keep an eye on that :)

**Comment by Zulex** [comment]
@cloudsyjared Cool project, will it only work with smartthings or are you planning make apis/libraries?

I had a look in the API of easybulb, won't that be enough to contol it by network? The server is given and authentication works with mac adresses, so we only need automatic detection of these adresses. The API can be found here : http://easybulb.com/api/

**Comment by cloudsyjared** [comment]
It's an exposed web api, so anyone can use it as long as they write the app to call it. Stay tuned
